[Security] #25091 add target user to SwitchUserListener

jwmickey opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #25091
| License       | MIT
| Doc PR        | 

This patch provides the target user to the SwitchUserListener's
accessDecisionManager->decide() call as the $object parameter to
give any registered voters extra information.

jwmickey committed:
[Security] #25091 add user to SwitchUserListener

This patch provides the target user to the SwitchUserListener's
accessDecisionManager->decide() call as the $object parameter to
give any registered voters extra information.

iltar commented on CHANGELOG-4.0.md:
```diff
@@ -1,6 +1,8 @@
 ===================
 
+* feature #25091 [Security] Add target user to SwitchUserListener (jwmickey)
+
```
No need to add this, will be done automatically when the version is released after merging.

xabbuh replied:
@iltar is right. But you need to add the feature to the component's changelog instead.

jwmickey replied:
I will move the changelog note to the correct file, thank you.

iltar commented on src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php:
```diff
@@ -126,18 +126,17 @@ private function attemptSwitchUser(Request $request, $username)
         }
 
         if (null !== $this->logger) {
             $this->logger->info('Attempting to switch to user.', array('username' => $username));
```
Won't this give the wrong username as result now?

xabbuh replied:
I wonder why we would need to do the change on line 129 instead.

jwmickey replied:
I don't think it will give the wrong username, but I did notice that I removed the line that sets attributes on the exception, so I'll add that back in.

jwmickey replied:
@iltar, my apologies, the code I submitted did indeed change the username for JSON requests.  The problem stemmed from me originally starting from the 2.7 branch and then rebasing to master, and I ended up changing some code from 2.7 that was different in master without realizing... in any case, lesson learned and thanks for pointing this out

jwmickey committed:
pull-request reworks

- moved message from the general changelog to the security component's
changelog
- restored exception attributes that were mistakenly removed

jwmickey committed:
fixed issue where username would be overwritten

xabbuh commented:
Can you please rebase on current `master`? Tests are currently failing.

jwmickey commented:
Is there anything else I need to do for this PR?

xabbuh commented on src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php:
```diff
@@ -126,18 +126,18 @@ private function attemptSwitchUser(Request $request, $username)
+        if (false === $this->accessDecisionManager->decide($token, array($this->role), $user)) {
             $exception = new AccessDeniedException();
             $exception->setAttributes($this->role);
-
```
this change should be reverted

jwmickey replied:
Why should this change be reverted?

jwmickey replied:
I'm sorry, I don't follow.  Are you saying that the $user param does not need to be passed to the accessDecisionManager->decide() call as the 3rd parameter?

xabbuh replied:
Oh that's a misunderstanding then. I was just referring to the blank line which should not have been removed.

xabbuh reviewed (approved):
apart from my minor comment

jwmickey committed:
Re-added empty line before throw statement

jwmickey committed:
Update SwitchUserListener.php

chalasr commented:
Thank you @jwmickey and congratz for your first contribution on Symfony!

jwmickey commented:
Whoohoo!  Thank you @chalasr, very exciting!
